samza-core/src/main/java/org/apache/samza/processorgraph/ProcessorGraph.java

```diff
@@ -0,0 +1,299 @@
+    this.config = config;
+  }
+
+  void addSource(StreamSpec input, String targetProcessorId) {
+    ProcessorNode node = getNode(targetProcessorId);
+    StreamEdge edge = getEdge(input);
+    edge.addTargetNode(node);
+    node.addInEdge(edge);
+    sources.add(edge);
+  }
+
+  void addSink(StreamSpec output, String sourceProcessorId) {
+    ProcessorNode node = getNode(sourceProcessorId);
+    StreamEdge edge = getEdge(output);
+    edge.addSourceNode(node);
+    node.addOutEdge(edge);
+    sinks.add(edge);
+  }
+
+  void addEdge(StreamSpec streamSpec, String sourceProcessorId, String targetProcessorId) {
+    ProcessorNode sourceNode = getNode(sourceProcessorId);
+    ProcessorNode targetNode = getNode(targetProcessorId);
+    StreamEdge edge = getEdge(streamSpec);
+    edge.addSourceNode(sourceNode);
+    edge.addTargetNode(targetNode);
+    sourceNode.addOutEdge(edge);
+    targetNode.addInEdge(edge);
+    internalStreams.add(edge);
+  }
+
+  ProcessorNode getNode(String processorId) {
+    ProcessorNode node = nodes.get(processorId);
+    if (node == null) {
+      node = new ProcessorNode(processorId, config);
+      nodes.put(processorId, node);
+    }
+    return node;
+  }
+
+  StreamEdge getEdge(StreamSpec streamSpec) {
+    String streamId = streamSpec.getId();
+    StreamEdge edge = edges.get(streamId);
+    if (edge == null) {
+      edge = new StreamEdge(streamSpec, config);
+      edges.put(streamId, edge);
+    }
+    return edge;
+  }
+
+  /**
+   * Returns the processor with configs to be executed in the topological order
+   * @return list of ProcessorConfig
+   */
+  public List<ProcessorNode> getProcessors() {
+    List<ProcessorNode> sortedNodes = topologicalSort();
+    return Collections.unmodifiableList(sortedNodes);
+  }
+
+  public Set<StreamEdge> getSources() {
+    return Collections.unmodifiableSet(sources);
+  }
+
+  public Set<StreamEdge> getSinks() {
+    return Collections.unmodifiableSet(sinks);
+  }
+
+  public Set<StreamEdge> getInternalStreams() {
+    return Collections.unmodifiableSet(internalStreams);
+  }
+
+
+  /**
+   * Validate the graph
+   */
+  public void validate() {
+    validateSources();
+    validateSinks();
+    validateInternalStreams();
+    validateReachability();
+  }
+
+  /**
+   * Validate the sources should have indegree being 0 and outdegree greater than 0
+   */
+  private void validateSources() {
+    sources.forEach(edge -> {
+        if (!edge.getSourceNodes().isEmpty()) {
+          throw new IllegalArgumentException(
+              String.format("Source stream %s should not have producers.", edge.getFormattedSystemStream()));
+        }
+        if (edge.getTargetNodes().isEmpty()) {
+          throw new IllegalArgumentException(
+              String.format("Source stream %s should have consumers.", edge.getFormattedSystemStream()));
+        }
+      });
+  }
+
+  /**
+   * Validate the sinks should have outdegree being 0 and indegree greater than 0
```

Comment:
Don't think it's necessary to make them intermediate streams. We can use the MessageStream that's the input to the sendTo as an input to the next operator too.